### DISCUSSION TITLE
Cannot find mpi.h during the ninja installation

### DISCUSSION THREAD

**Discussion by HANSIeltsKing** [su2code/SU2#1692] [Q&A]
Hi, I want to build SU2 from source using meson and ninja. After I enabled mpi and tried to install by typing "./ninja -C build install" an error occurs:

```
......
[14/1296] Compiling C object externals/cgns/hdf5/libsu2hdf5.a.p/H5Aint.c.o
FAILED: externals/cgns/hdf5/libsu2hdf5.a.p/H5Aint.c.o
cc -Iexternals/cgns/hdf5/libsu2hdf5.a.p -Iexternals/cgns/hdf5 -I../externals/cgns/hdf5 -fdiagnostics-color=always -g -fPIC -Wno-unused-result -Wno-unused-parameter -Wno-unused-variable -Wno-unused-but-set-variable -Wno-sign-compare -Wno-stringop-overflow -Wno-discarded-qualifiers -Wno-error=maybe-uninitialized -Wno-error=unused-function -Wno-error=unused-label -Wno-pedantic -D_FILE_OFFSET_BITS=64 -D_LARGEFILE64_SOURCE -D_LARGEFILE_SOURCE -D_POSIX_C_SOURCE=200809L -D_GNU_SOURCE -MD -MQ externals/cgns/hdf5/libsu2hdf5.a.p/H5Aint.c.o -MF externals/cgns/hdf5/libsu2hdf5.a.p/H5Aint.c.o.d -o externals/cgns/hdf5/libsu2hdf5.a.p/H5Aint.c.o -c ../externals/cgns/hdf5/H5Aint.c
In file included from ../externals/cgns/hdf5/H5private.h:27,
                 from ../externals/cgns/hdf5/H5Aint.c:35:
../externals/cgns/hdf5/H5public.h:63:10: fatal error: mpi.h: No such file or directory
   63 | #include <mpi.h>
      |          ^~~~~~~
compilation terminated.
ninja: build stopped: subcommand failed.
```
All the error output is due to this problem of missing mpi.h, but I can confirm that mpi.h is in that directory.

**Comment by martin-g** [comment]
Hi,

> After I enabled mpi 

What exactly did you do ?

>  but I can confirm that mpi.h is in that directory

It is not clear in which directory ?

On Linux Ubuntu you may install `libopenmpi-dev` package. It will provide `/usr/lib/x86_64-linux-gnu/openmpi/include/mpi.h`

**Reply by HANSIeltsKing** [reply]
I did it by -Dcustom-mpi=true

I can find a mpi.h under the openmpi
However, I do not know what is the location does ninja install.
How can I solve this problem? I want to install an SU2 from source with custom-mpi

**Comment by martin-g** [comment]
Please also read https://su2code.github.io/docs_v7/Build-SU2-Linux-MacOS/ for more information.

**Reply by HANSIeltsKing** [reply]
Thank you very much!

**Comment by jjywow** [comment]
Hello,

I encountered exact same error.
"../externals/cgns/hdf5/H5public.h:63:10: fatal error: mpi.h: No such file or directory
   63 | #include <mpi.h>
      |          ^~~~~~~
compilation terminated.
ninja: build stopped: subcommand failed."

OS: WSL ubuntu-20.04

I compile meson by following
./meson.py build -Dcustom-mpi=true -Denable-pywrapper=true

I have mpi.h file following directory wsl$\Ubuntu-20.04\usr\lib\x86_64-linux-gnu\openmpi\include
I installed mpi4py following directory wsl$\Ubuntu-20.04\usr\lib\python3\dist-packages\mpi4py\include\mpi4py

i installed pkg-config and add path by
export CC=\usr\bin\mpicc
export CXX=\usr\bin\mpicxx

I have SU2 directory as follows \wsl$\Ubuntu-20.04\home\jeeyoungek\SU2

I really appreciate anyone's help

**Reply by bigfooted** [reply]
How did you install mpi? Is the folder that contains mpi.h included in the search path, usually LD_LIBRARY_PATH? This should be done automatically if you install using a package manager, e.g. 
> sudo apt install libopenmpi-dev

**Reply by jjywow** [reply]
Hello, thanks for reply.

I did use sudo apt install libopenmpi-dev for mpi

but I have no idea about 'LD_LIBRARY_PATH'

I tried echo $LD_LIBRARY_PATH but responed nothing

so i added  by 'export LD_LIBRARY_PATH=/usr/lib/local/x86_64-linux-gnu/openmpi/include:$LD_LIBRARY_PATH'

and tried to compile again but same error. could you provide more suggestions?

**Reply by bigfooted** [reply]
Can you check that when you run meson, that the compiler and linker use mpicc and mpicxx and not the regular compiler/linker?

**Reply by jjywow** [reply]
Hello,
below is the results when I run meson.
I don't know much about compiler and linker,  I guess gcc is used from the below

```
jeeyoungek@DESKTOP-jjy115:~/SU2$ ./meson.py build --reconfigure -Dcustom-mpi=true -Denable-pywrapper=true
The Meson build system
Version: 0.61.1
Source dir: /home/jeeyoungek/SU2
Build dir: /home/jeeyoungek/SU2/build
Build type: native build
Project name: SU2
Project version: 7.4.0 "Blackbird"
C compiler for the host machine: cc (gcc 9.4.0 "cc (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0")
C linker for the host machine: cc ld.bfd 2.34
C++ compiler for the host machine: c++ (gcc 9.4.0 "c++ (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0")
C++ linker for the host machine: c++ ld.bfd 2.34
Host machine cpu family: x86_64
Host machine cpu: x86_64
Program python3 found: YES (/usr/bin/python3)

```

**Reply by bigfooted** [reply]
looks like it uses the regular gcc and not mpicc. Do you need the flag -Dcustom-mpi=true? What happens if you leave it out? I think it is always necessary to define the MPICC and MPICXX paths if you include custom-mpi

**Reply by jjywow** [reply]
Hello, 
Yes, it would be better for me to use mpi. 
Thanks to your comment, I searched for defining MPICC and MPICXX. 
I properly redefine the MPICC and MPICXX, and it worked. 
Thank you very much

**Comment by jjywow** [comment]
Below is pretty much I did

python 3.8
sudo apt update
sudo apt install python3.8 –y

gcc
sudo apt update
sudo apt install build-essential

mpi
sudo apt install libopenmpi-dev

pkg-config
apt install pkg-config

numpy, scipy
sudo apt install python3-numpy

brew
sudo apt-get install build-essential curl file git
sh -c "$(curl -fsSL https://raw.githubusercontent.com/Linuxbrew/install/master/install.sh)"
export PATH="/home/linuxbrew/.linuxbrew/bin:$PATH"
export MANPATH="/home/linuxbrew/.linuxbrew/share/man:$MANPATH"
export INFOPATH="/home/linuxbrew/.linuxbrew/share/info:$INFOPATH"

swig install
brew install swig

pip install 
sudo apt-get install python3-pip 
sudo pip install --upgrade pip 
sudo pip install --upgrade virtualenv

mpi4py
pip install mpi4py

export PATH="/home/linuxbrew/.linuxbrew/bin:$PATH"
export PATH="/usr/lib/x86_64-linux-gnu/openmpi/include:$PATH"
export CC=/usr/bin/mpicc
export CXX=/usr/bin/mpicxx
export LD_LIBRARY_PATH=/usr/lib/local/x86_64-linux-gnu/openmpi/include:$LD_LIBRARY_PATH

./meson.py build -Dcustom-mpi=true -Denable-pywrapper=true
./ninja -C build install
